monoids.md

```diff
@@ -0,0 +1,413 @@
+mempty :: Min Int
+> Min {getMin = 9223372036854775807}
+mempty :: Max Int
+> Max {getMax = -9223372036854775808}
+```
+
+This means that if we fold an empty container, we get a value which is confusing in some cases.
+
+```haskell
+fold [] :: Min Int
+> Min {getMin = 9223372036854775807}
+```
+
+Instead we sometimes want to represent this "failure" using Maybe.
+
+
+## Lifting a Semigroup into a Monoid
+
+If we take a look at the Maybe data type : `data Maybe a = Nothing | Just a`.
+
+We can see that it can represent a set of values (`Just a`) with an additional element `Nothing`.
+
+If the `a`'s are monoids, we can combine them using `<>` and then chose `Nothing` as our identity element.
+In other words, we embeded a semigroup in a monoid by adjoining an identity.
+
+The monoid instance for `Maybe` implements this idea:
+
+```haskell
+
+instance Semigroup a => Semigroup (Maybe a) where
+    Nothing <> b       = b
+    a       <> Nothing = a
+    Just a  <> Just b  = Just (a <> b)
+
+instance Semigroup a => Monoid (Maybe a) where
+    mempty = Nothing
+
+fold [Just (Sum 1), Just (Sum 2),Nothing]
+> Just (Sum 3)
+```
+
+We can use this to fold into a semigroup.
+An example of a semigroup that is not a monoid is `First`, which has the semigroup implementation `a <> _ = a`, meaning that it will always pick the left argument.
+There is understandably no identity element for this semigroup,
+but by using `Maybe (First Int)` we use `Nothing` as our identity.
+We have now created a monoid from the First semigroup, and can therefore use it in a fold.
+
+```haskell
+foldMap (Just . First) [1,2,3]
+> Just (First 1)
+```
+
+Back to our "problem" with the min/max identity elements:
+If we now wrap up our type inside a `Maybe`, the result is more sensible(?)
+```haskell
+>fold [] :: Maybe (Min Int)
+Nothing
+```
+
+
+## Lists
+
+Lists are monoids with `(<>) = (++)` and `mempty = []`.
+
+```haskell
+[1,2,3] <> [4,5]
+> [1,2,3,4,5]
+
+foldMap (\x -> [x,x*2]) [1,2,3]
+> [1,2,2,4,3,6]
+
+```
+
+In fact, the list is an important monoid, called the free monoid.
+This comes from the fact that if you take the requirements for a monoid and create the minimal strcuture needed to satisfy these requirements, 
+you end up with list monoid instance.
+
+
+### Tuples
+A very nice neat property about monoids is that they compose over products.
+Which in practice means, that if you have two monoids, you can put combine them in a tuple and then that tuple is also a monoid.
+
+```haskell
+instance (Monoid a, Monoid b) => Monoid (a,b) where
+    (a,b) <> (a',b') = (a<>a',b<>b')
+    
+instance (Monoid a, Monoid b) => Monoid (a,b) where
+        mempty = (mempty, mempty)
+
+("abc",Sum 2) <> ("abc",Sum 1)
+> ("abcabc", Sum 3)
+```
+
+(This also implemented for tuples of size 3,4 and 5)
+
+This makes it easy to accumulate multiple monoid values in a single fold.
+We can use this to find the minimum,maximum,length,sum and product of a list, all in one pass in a neat way. 
```

Comment:
Fint med litt mellomrom. Grammarly ville ha en komma før and her også. Jeg stoler på grammarly 😄 

```suggestion
We can use this to find the minimum, maximum, length, sum, and product of a list, all in one pass in a neat way. 
```

Reply:
Hmm, er lov med begge.

https://www.grammarly.com/blog/comma-before-and/#:~:text=Whether%20or%20not%20you%20put,before%20and%20in%20a%20list.